chore: Update Sass Library

noahtallen opened the pull request:
#### Changes proposed in this Pull Request
1. Update dependency `sass` to 1.37.5 and `sass-loader` to 12.1.0.
2. Run `sass-migrator division` across the codebase
3. Change our webpack sass prelude to use `@use` instead of `@import`
4. Hide deprecation warnings from dependencies.

This PR is motivated by #54793. Specifically, the new wp monorepo update also updates the version of sass. The new version of sass includes a deprecation warning for the division operator (`/`). [More info on that here](https://sass-lang.com/documentation/breaking-changes/slash-div). This deprecation warning prints thousands  of times during the build process and I even got a segfault running the build which miiiight be related. It also fills the TeamCity build log until it reaches the log limit, so we can't see any actual problems with the build, just the dep warning. As a result, we can't really merge that update without resolving this deprecation warning.

```
DEPRECATION WARNING: Using / for division is deprecated and will be removed in Dart Sass 2.0.0.

Recommendation: math.div($pixels, $context)

More info and automated migrator: https://sass-lang.com/d/slash-div
```

Thankfully, sass has a [migrator tool](https://github.com/sass/migrator#readme) which lets us auto-fix this problem.

However, that raises a different problem, which is that all `@use` directives must be the very first thing in the file. The new division operator is imported via `@use`:

```
ERROR in ../packages/design-picker/src/components/style.scss
Module build failed (from ../node_modules/mini-css-extract-plugin/dist/loader.js):
ModuleBuildError: Module build failed (from ../node_modules/sass-loader/dist/cjs.js):
SassError: @use rules must be written before any other rules.
   |
 2 │ @use 'sass:math';
   | ^^^^^^^^^^^^^^^^
   |
 packages/design-picker/src/components/style.scss 2:1  root stylesheet
```

I discovered the culprit, which is our `prelude` (`additionalData`) which is added to every sass file in the project. It basically means that an `import` statement is inserted as first thing in every file instead of the `use` statement. Thankfully, this can be migrated to the new format like so:

```diff
- @import 'path/to/_utils.scss';
+ @use 'path/to/_utils.scss' as *;
```

The `as *` defines the namespace for the import. Without it, you'd have to start referencing variables with the namespace prefixed. While that sounds like a good idea to implement in the long run, we have hundreds of references to these variables which assume they are defined globally. As such, we need to continue defining them globally with the `*` namespace.

Finally, I also needed to set the `quietDeps` flag in our sass options. This flag disables deprecation warnings when processing imported/3rd-party stylesheets. This is important, as we can't fix those deprecation warnings ourselves, and it adds so much noise to the build log that TeamCity can't display the entire logfile.

#### Testing instructions
- Smoke test stylesheets across the codebase
- CI should pass

noahtallen committed:
Update sass and sass-loader

noahtallen committed:
Run 'sass-migrator division' to fix warning

noahtallen committed:
Use @use instead of @import for prelude

noahtallen committed:
Hide dep warnings for dependencies

noahtallen committed:
Deduplicate dependencies

noahtallen committed:
Revert sass update in certain packages

noahtallen commented:
Currently getting a lot of `[webpack.cache.PackFileCacheStrategy] Error: No serializer registered for SassError` errors, which I'll try to figure out tomorrow. @scinos do you know anything about that error?

scinos commented:
`[webpack.cache.PackFileCacheStrategy] Error: No serializer registered for SassError` is a warning that says that particular module won't be cached because `sass-loader` doesn't provide a serializer for SassError. It makes the log so noisy that it hides the actual error (`#15 199.6 [output clipped, log limit 1MiB reached]`).

When building this branch locally with `PROFILE=true PERSISTENT_CACHE=true yarn build-client-evergreen` I get a ton of error like this one:

```
ERROR in ../apps/notifications/src/panel/boot/stylesheets/style.scss
Module build failed (from ../node_modules/mini-css-extract-plugin/dist/loader.js):
ModuleBuildError: Module build failed (from ../packages/calypso-build/node_modules/sass-loader/dist/cjs.js):
SassError: Undefined function.
   ╷
28 │     @return math.div($pixels, $context) * 1rem;
   │             ^^^^^^^^^^^^^^^^^^^^^^^^^^^
   ╵
  packages/typography/styles/variables.scss 28:10               rem()
  packages/typography/styles/variables.scss 38:23               @import
  packages/typography/styles/fonts.scss 1:9                     @import
  client/assets/stylesheets/shared/_typography.scss 1:9         @import
  client/assets/stylesheets/shared/_utils.scss 4:9              @use
  apps/notifications/src/panel/boot/stylesheets/style.scss 1:1  root stylesheet
    at processResult (/Users/sergio/src/automattic/wp-calypso/node_modules/webpack/lib/NormalModule.js:703:19)
    at /Users/sergio/src/automattic/wp-calypso/node_modules/webpack/lib/NormalModule.js:809:5
    at /Users/sergio/src/automattic/wp-calypso/node_modules/loader-runner/lib/LoaderRunner.js:399:11
    at /Users/sergio/src/automattic/wp-calypso/node_modules/loader-runner/lib/LoaderRunner.js:251:18
    at context.callback (/Users/sergio/src/automattic/wp-calypso/node_modules/loader-runner/lib/LoaderRunner.js:124:13)
    at /Users/sergio/src/automattic/wp-calypso/packages/calypso-build/node_modules/sass-loader/dist/index.js:54:7
    at Function.call$2 (/Users/sergio/src/automattic/wp-calypso/node_modules/sass/sass.dart.js:92646:16)
    at _render_closure1.call$2 (/Users/sergio/src/automattic/wp-calypso/node_modules/sass/sass.dart.js:81147:12)
    at _RootZone.runBinary$3$3 (/Users/sergio/src/automattic/wp-calypso/node_modules/sass/sass.dart.js:27268:18)
    at _FutureListener.handleError$1 (/Users/sergio/src/automattic/wp-calypso/node_modules/sass/sass.dart.js:25818:19)
 @ ../apps/notifications/src/panel/Notifications.jsx 13:0-39
 @ ./notifications/index.jsx 16:0-101 106:17-29 164:6-18 244:40-58
 @ ./layout/masterbar/notifications.jsx 17:2-19:26
 @ ./layout/masterbar/logged-in.jsx 37:0-44 312:43-56
 @ ./layout/index.jsx 24:0-67 298:91-108
 @ ./controller/index.web.js 14:0-36 41:65-71
 @ ./boot/common.js 11:0-59 393:57-78
 @ ./boot/app.js 3:0-35 7:2-9
```

scinos committed:
Import shared utils in using * namespace

noahtallen commented:
> The prelude in this branch doesn't include the as * part. When I added it, it fixed the above errors.

Ahhh, I must have accidentally removed it when I rebased.
